### PR TITLE
carli/1867_large_region_select_performance_improvement

### LINKED ISSUE #1867: performance impact to the file browser when trying to export regions from a large set of regions
Moving the file browser becomes much less responsive when the region list is long. 

To experience this issue, please try to load 10 regions and launch the file browser for region exporting. Then try to drag and drop the file browser to a new location (or try to resize the dialog). This should give us a smooth user experience. Now repeat the exercise by loading 1000 regions instead. The file browser becomes much less responsive when we try to resize the dialog or relocate the dialog. 

https://user-images.githubusercontent.com/20819712/169963478-7e5da254-fafd-4d1a-b2ae-275443b7e499.mov

### PR DESCRIPTION
**Pull request by crocka**
**Description**
This PR closes #1867. It uses ``react-window`` to create a virtualized list of region that are only rendered when the users scroll to view. 

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] e2e test passing / ~corresponding fix added~ / ~new e2e test created~
- [x] changelog updated / no changelog update needed
- [x] protobuf updated to the latest dev commit / no protobuf update needed
- [x] protobuf version bumped / protobuf version not bumped
- [x] `BackendService` unchanged / `BackendService` changed and corresponding ICD test fix added

### CODE REVIEW

**Review by kswang1029 — APPROVED** [review]
Looks good. No regression from e2e tests. Please proceed code review.

**Review by kswang1029 — APPROVED** [review]
retested. The browser is responsive when there are 4000 regions in my test case. Selecting a region can also trigger a region highlight in the region list of the file browser. No regression from e2e tests.

File: package.json [hunk 1]
```diff
@@ -203,4 +203,4 @@
         "react-scripts": "^5.0.1",
         "usehooks-ts": "^2.9.1"
     }
-}
+}
```
Comment (YuHsuan-Hwang) [inline package.json:206]: Remove changes in `package.json` and `package-lock.json`.

File: src/components/Dialogs/FileBrowser/RegionSelect/RegionSelectComponent.tsx [hunk 2]
```diff
@@ -10,6 +12,8 @@ import "./RegionSelectComponent.scss";
 
 @observer
 export class RegionSelectComponent extends React.Component {
+    private reference: HTMLDivElement;
```
Comment (YuHsuan-Hwang) [inline src/components/Dialogs/FileBrowser/RegionSelect/RegionSelectComponent.tsx]: Is this unused?

File: src/stores/FileBrowserStore/FileBrowserStore.ts [hunk 3]
```diff
@@ -536,15 +536,17 @@ export class FileBrowserStore {
 
     @action addExportRegionIndex = (regionIndex: number) => {
         if (!this.exportRegionIndexes.includes(regionIndex)) {
-            this.exportRegionIndexes.push(regionIndex);
-            this.exportRegionIndexes.sort();
+            this.exportRegionIndexes = [...this.exportRegionIndexes, regionIndex];
+            this.exportRegionIndexes = [...this.exportRegionIndexes].sort();
```
Comment (YuHsuan-Hwang) [inline src/stores/FileBrowserStore/FileBrowserStore.ts]: 1. I think we only need to copy the array once within the action function.
2. Should `sort()` be changed to `sort((a, b) => a - b)`?

**Review by YuHsuan-Hwang — CHANGES_REQUESTED** [review]
The feature looks good. 👍 I have some minor comments.

**Comment by loveluthien** [comment]
The export part is good, but the import part has a similar issue. The window of the import region becomes less responsive when we click a file with lots of regions. Do we want to deal with the import part in this PR?

**Comment by YuHsuan-Hwang** [comment]
> The export part is good, but the import part has a similar issue. The window of the import region becomes less responsive when we click a file with lots of regions. Do we want to deal with the import part in this PR?

Let's work on it with a separate PR.
@loveluthien Could you help file a new issue for this?

**Review by loveluthien — APPROVED** [review]
It looks good. I will file issue about the import part.

**Review by YuHsuan-Hwang — APPROVED** [review]

**Pull request merged**